### PR TITLE
Select Go version from go.mod

### PR DESCRIPTION
**Pull request by radiosilence**
* A short explanation of the proposed change:

I have written an extremely simple parser that finds the Go version in go.mod and uses that if possible.

Ideally we'd use the real parser here https://pkg.go.dev/golang.org/x/mod@v0.5.1/modfile#Parse but from what I can see it is not in stable and also possibly internal.

* An explanation of the use cases your change solves:

Go module was supported but still needed a GOVERSION env override to select correct version, which was confusing to me as someone new to Cloud Foundry

Here is an example repo which works with Go 1.18 features without need for GOVERSION env: https://github.com/radiosilence/hello-go-cloud-foundry-generics

* [x] I have viewed signed and have submitted the Contributor License Agreement

* [x] I have made this pull request to the `develop` branch

* [x] I have added an integration test *I am not quite sure how to get Machete running yet - I have added a sane seeming test though*

### CODE REVIEW

**Comment by radiosilence** [comment]
Can a maintainer approve running the workflow?

**Comment by brayanhenao** [comment]
Hi @radiosilence, thanks for your PR!

Could you fix the Unit tests and also add an integration test case that covers this scenario? Thanks!

**Comment by radiosilence** [comment]
Hi @brayanhenao just trying to get the unit tests running and I get this

```
! ./scripts/unit.sh

Installing ginkgo
go: go.mod file not found in current directory or any parent directory.
        'go get' is no longer supported outside a module.
        To build and install a command, use 'go install' with a version,
        like 'go install example.com/cmd@latest'
        For more information, see https://golang.org/doc/go-get-install-deprecation
        or run 'go help get' or 'go help install'.
```

Will I have to use an older go version?

**Comment by radiosilence** [comment]
Hi @brayanhenao  I've fixed the unit tests but getting the integration tests running is proving frustrating on an M1 mac

File: src/go/gomod/gomod.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	"os"
+	"strings"
+)
+
+type GoMod struct {
+	GoVersion string
+}
+
+// TODO: Replace with https://pkg.go.dev/golang.org/x/mod/modfile when it is available
+func (g *GoMod) Load(path string) {
```
Comment (brayanhenao) [inline src/go/gomod/gomod.go:15]: This function should return an error. Also, every function used inside that may generate an error should be returned with a clear error message, do not use `log.fatal`.

Something like this:

```go
func (g *GoMod) Load(path string) error {
	f, err := os.Open(path)
	if err != nil {
		return fmt.Errorf("failed to open go.mod: %w", err)
	}
	.
	.
	.
	return nil
}
```

File: src/go/supply/supply.go [hunk 2]
```diff
@@ -93,9 +95,13 @@ func (gs *Supplier) SelectVendorTool() error {
...
 
-	if exists, err := libbuildpack.FileExists(filepath.Join(gs.Stager.BuildDir(), "go.mod")); err != nil {
+	isGoMod, err := libbuildpack.FileExists(filepath.Join(gs.Stager.BuildDir(), "go.mod"))
+	if err != nil {
 		return err
-	} else if exists {
+	}
+	if isGoMod {
+		gs.Log.BeginStep("Checking go.mod file")
+		gs.GoMod.Load(filepath.Join(gs.Stager.BuildDir(), "go.mod"))
```
Comment (brayanhenao) [inline src/go/supply/supply.go:104]: And here should be done the handling of any error returned by `Load`.

**Review by brayanhenao — CHANGES_REQUESTED** [review]

File: src/go/gomod/gomod.go (hunk 1 above)
Comment (radiosilence) [reply src/go/gomod/gomod.go:15]: Ok, thanks for the help!

File: src/go/supply/supply.go (hunk 2 above)
Comment (radiosilence) [reply src/go/supply/supply.go:104]: Yep, makes sense. We should warn and not select the version.

**Review by brayanhenao — APPROVED** [review]

**Comment by radiosilence** [comment]
Hi @brayanhenao I see this was merged, and in 1.9.44, but then removed in 1.9.46. Any reason why this happened?

**Pull request merged**